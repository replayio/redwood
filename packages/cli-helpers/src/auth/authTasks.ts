import fs from 'fs'
import path from 'path'

import execa from 'execa'
import { ListrTask, ListrTaskWrapper, ListrRenderer } from 'listr2'

import { writeFilesTask, transformTSToJS } from '../lib'
import { colors } from '../lib/colors'
import { getPaths, resolveFile } from '../lib/paths'
import {
  isTypeScriptProject,
  getGraphqlPath,
  graphFunctionDoesExist,
} from '../lib/project'

import { apiSideFiles, generateUniqueFileNames } from './authFiles'

const AUTH_PROVIDER_HOOK_IMPORT = `import { AuthProvider, useAuth } from './auth'`
const AUTH_HOOK_IMPORT = `import { useAuth } from './auth'`

export const getWebAppPath = () => getPaths().web.app

const addApiConfig = (authDecoderImport?: string) => {
  const graphqlPath = getGraphqlPath()

  if (!graphqlPath) {
    throw new Error('Could not find your graphql file path')
  }

  let content = fs.readFileSync(graphqlPath).toString()
  let contentUpdated = false

  if (authDecoderImport && !content.includes(authDecoderImport)) {
    content = authDecoderImport + '\n' + content

    // If we have multiple auth providers setup we probably already have an
    // auth decoder configured. In that case we don't want to add another one
    // Have to use a funky looking Regex here to prevent a "Polynomial regular
    // expression used on uncontrolled data" warning/error. A.k.a "Catastrophic
    // Backtracking". The usual fix is to use an atomic group, but the JS
    // regex engine doesn't support that, so we use a lookaround group to
    // emulate an atomic group.
    if (
      !new RegExp(
        '(?=(^.*?createGraphQLHandler))\\1.*\\bauthDecoder',
        's'
      ).test(content)
    ) {
      content = content.replace(
        /^(\s*)(loggerConfig:)(.*)$/m,
        `$1authDecoder,\n$1$2$3`
      )
    }

    contentUpdated = true
  }

  const hasAuthImport =
    /(^import {.*?getCurrentUser(?!getCurrentUser).*?} from 'src\/lib\/auth')/s.test(
      content
    )

  if (!hasAuthImport) {
    // add import statement
    content = content.replace(
      /^(import { db } from 'src\/lib\/db')$/m,
      `import { getCurrentUser } from 'src/lib/auth'\n$1`
    )

    // add object to handler
    content = content.replace(
      /^(\s*)(loggerConfig:)(.*)$/m,
      `$1getCurrentUser,\n$1$2$3`
    )

    contentUpdated = true
  }

  if (contentUpdated) {
    fs.writeFileSync(graphqlPath, content)
  }
}

const apiSrcDoesExist = () => {
  return fs.existsSync(path.join(getPaths().api.src))
}

const webIndexDoesExist = () => {
  return fs.existsSync(getWebAppPath())
}

const addAuthImportToApp = (content: string) => {
  const contentLines = content.split('\n')
  // Find the last import line that's not a .css or .scss import
  // @ts-expect-error - We polyfill findLastIndex
  const importIndex = contentLines.findLastIndex((line: string) =>
    /^\s*import (?!.*(?:.css'|.scss'))/.test(line)
  )

  // After the import found above, insert a blank line followed by the
  // AuthProvider and useAuth import
  contentLines.splice(importIndex + 1, 0, '', AUTH_PROVIDER_HOOK_IMPORT)

  return contentLines.join('\n')
}

const addAuthImportToRoutes = (content: string) => {
  const contentLines = content.split('\n')
  // Find the last import line that's not a .css or .scss import
  // @ts-expect-error - We polyfill findLastIndex
  const importIndex = contentLines.findLastIndex((line: string) =>
    /^\s*import (?!.*(?:.css'|.scss'))/.test(line)
  )

  // After the import found above, insert a blank line followed by the
  // useAuth import
  contentLines.splice(importIndex + 1, 0, '', AUTH_HOOK_IMPORT)

  return contentLines.join('\n')
}

const hasAuthProvider = (content: string) => {
  return /\s*<AuthProvider>/.test(content)
}

/** returns the content of App.{js,tsx} with <AuthProvider> added */
const addAuthProviderToApp = (content: string) => {
  const match = content.match(
    /(\s+)(<RedwoodProvider.*?>)(.*)(<\/RedwoodProvider>)/s
  )

  if (!match) {
    throw new Error('Could not find <RedwoodProvider> in App.{js,tsx}')
  }

  const [
    _,
    newlineAndIndent,
    redwoodProviderOpen,
    redwoodProviderChildren,
    redwoodProviderClose,
  ] = match

  const redwoodProviderChildrenLines = redwoodProviderChildren
    .split('\n')
    .map((line, index) => {
      return `${index === 0 ? '' : '  '}` + line
    })

  const renderContent =
    newlineAndIndent +
    redwoodProviderOpen +
    newlineAndIndent +
    '  ' +
    `<AuthProvider>` +
    redwoodProviderChildrenLines.join('\n') +
    `</AuthProvider>` +
    newlineAndIndent +
    redwoodProviderClose

  return content.replace(
    /\s+<RedwoodProvider.*?>.*<\/RedwoodProvider>/s,
    renderContent
  )
}

const hasUseAuthHook = (componentName: string, content: string) => {
  return new RegExp(
    `<${componentName}.*useAuth={.*?}.*?>.*<\/${componentName}>`,
    's'
  ).test(content)
}

const addUseAuthHook = (componentName: string, content: string) => {
  return content.replace(
    `<${componentName}`,
    `<${componentName} useAuth={useAuth}`
  )
}

/**
 * Actually inserts the required config lines into App.{js,tsx}
 * Exported for testing
 */
export const addConfigToApp = async () => {
  const webAppPath = getWebAppPath()

  let content = fs.readFileSync(webAppPath).toString()

  if (!content.includes(AUTH_PROVIDER_HOOK_IMPORT)) {
    content = addAuthImportToApp(content)
  }

  if (!hasAuthProvider(content)) {
    content = addAuthProviderToApp(content)
  }

  if (/\s*<RedwoodApolloProvider/.test(content)) {
    if (!hasUseAuthHook('RedwoodApolloProvider', content)) {
      content = addUseAuthHook('RedwoodApolloProvider', content)
    }
  } else {
    console.warn(
      colors.warning(
        'Could not find <RedwoodApolloProvider>.\nIf you are using a custom ' +
          'GraphQL Client you will have to make sure it gets access to your ' +
          '`useAuth`, if it needs it.'
      )
    )
  }

  fs.writeFileSync(webAppPath, content)
}

export const createWebAuth = (
  basedir: string,
  provider: string,
  webAuthn: boolean
) => {
  const templatesBaseDir = path.join(basedir, 'templates', 'web')
  const templates = fs.readdirSync(templatesBaseDir)

  const templateFileName = templates.find((template) => {
    return template.startsWith('auth.' + (webAuthn ? 'webAuthn.ts' : 'ts'))
  })

  if (!templateFileName) {
    throw new Error('Could not find the auth.ts template')
  }

  const templateExtension = templateFileName.split('.').at(-2)

  // Find an unused filename
  // Start with web/src/auth.{ts,tsx}
  // Then web/src/providerAuth.{ts,tsx}
  // Then web/src/providerAuth2.{ts,tsx}
  // Then web/src/providerAuth3.{ts,tsx}
  // etc
  let authFileName = path.join(getPaths().web.src, 'auth')
  let i = 1
  while (resolveFile(authFileName)) {
    const count = i > 1 ? i : ''

    authFileName = path.join(getPaths().web.src, provider + 'Auth' + count)

    i++
  }

  authFileName = authFileName + '.' + templateExtension

  let template: string | undefined = fs.readFileSync(
    path.join(templatesBaseDir, templateFileName),
    'utf-8'
  )

  template = isTypeScriptProject()
    ? template
    : transformTSToJS(authFileName, template)

  fs.writeFileSync(authFileName, template)
}

export const addConfigToRoutes = () => {
  const webRoutesPath = getPaths().web.routes

  let content = fs.readFileSync(webRoutesPath).toString()

  if (!content.includes(AUTH_HOOK_IMPORT)) {
    content = addAuthImportToRoutes(content)
  }

  if (!hasUseAuthHook('Router', content)) {
    content = addUseAuthHook('Router', content)
  }

  fs.writeFileSync(webRoutesPath, content)
}

export const generateAuthApiFiles = <Renderer extends typeof ListrRenderer>(
  basedir: string,
  provider: string,
  force: boolean,
  webAuthn: boolean
): ListrTask<never, Renderer> => {
  return {
    title: 'Generating auth api side files...',
    task: (_ctx: never, task: ListrTaskWrapper<never, Renderer>) => {
      if (!apiSrcDoesExist()) {
        return task.skip?.('api/src not found, skipping')
      }

      // The keys in `filesRecord` are the full paths to where the file contents,
      // which is the values in `filesRecord`, will be written.
      let filesRecord = apiSideFiles({ basedir, webAuthn })

      if (!force) {
        const uniqueFilesRecord = generateUniqueFileNames(filesRecord, provider)

        if (
          Object.keys(filesRecord).join(',') !==
          Object.keys(uniqueFilesRecord).join(',')
        ) {
          console.warn(
            colors.warning(
              "To avoid overwriting existing files we've generated new file " +
                'names for the newly generated files. This probably means ' +
                `${provider} auth doesn't work out of the box. You'll most ` +
                'likely have to manually merge some of the generated files ' +
                'with your existing auth files'
            )
          )
        }

        filesRecord = uniqueFilesRecord
      }

      return writeFilesTask(filesRecord, { overwriteExisting: force })
    },
  }
}

export const addAuthConfigToWeb = <Renderer extends typeof ListrRenderer>(
  basedir: string,
  provider: string,
  webAuthn = false
) => ({
  title: 'Adding auth config to web...',
  task: (_ctx: never, task: ListrTaskWrapper<never, Renderer>) => {
    if (webIndexDoesExist()) {
      addConfigToApp()
      createWebAuth(basedir, provider, webAuthn)
      addConfigToRoutes()
    } else {
      task.skip?.(
        `web/src/App.${
          isTypeScriptProject() ? 'tsx' : 'js'
        } not found, skipping`
      )
    }
  },
})

export const addAuthConfigToGqlApi = <Renderer extends typeof ListrRenderer>(
  authDecoderImport?: string
) => ({
  title: 'Adding auth config to GraphQL API...',
  task: (_ctx: never, task: ListrTaskWrapper<never, Renderer>) => {
    if (graphFunctionDoesExist()) {
      addApiConfig(authDecoderImport)
    } else {
      task.skip?.('GraphQL function not found, skipping')
    }
  },
})

export const addWebPackages = (webPackages: string[], rwVersion: string) => ({
  title: 'Adding required web packages...',
  task: async () => {
    const args = [
      'workspace',
      'web',
      'add',
      ...webPackages,
      `@redwoodjs/auth@${rwVersion}`,
    ]
    await execa('yarn', args)
  },
})

export const addApiPackages = (apiPackages: string[]) =>
  apiPackages.length > 0 && {
    title: 'Adding required api packages...',
    task: async () => {
      await execa('yarn', ['workspace', 'api', 'add', ...apiPackages])
    },
  }

export const installPackages = {
  title: 'Installing packages...',
  task: async () => {
    await execa('yarn', ['install'])
  },
}

export const printNotes = <Renderer extends typeof ListrRenderer>(
  notes: string[]
) => ({
  title: 'One more thing...',
  task: (_ctx: never, task: ListrTaskWrapper<never, Renderer>) => {
    task.title = `One more thing...\n\n   ${notes.join('\n   ')}\n`
  },
})
