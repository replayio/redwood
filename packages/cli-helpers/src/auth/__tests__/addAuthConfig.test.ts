// Have to use `var` here to avoid "Temporal Dead Zone" issues
// eslint-disable-next-line
var mockWebAppPath = ''
// eslint-disable-next-line
var mockWebRoutesPath = ''

import fs from 'fs'
import path from 'path'

// mock Telemetry for CLI commands so they don't try to spawn a process
jest.mock('@redwoodjs/telemetry', () => {
  return {
    errorTelemetry: () => jest.fn(),
    timedTelemetry: () => jest.fn(),
  }
})

import { addConfigToApp, addConfigToRoutes, createWebAuth } from '../authTasks'

jest.mock('../../lib/paths', () => {
  const path = require('path')
  const __dirname = path.resolve()
  const originalModule = jest.requireActual('../../lib/paths')

  return {
    resolveFile: originalModule.resolveFile,
    getPaths: () => ({
      api: { functions: '', src: '', lib: '' },
      web: {
        src: path.join(__dirname, '../create-redwood-app/template/web/src'),
        app: path.join(
          __dirname,
          mockWebAppPath || '../create-redwood-app/template/web/src/App.tsx'
        ),
        routes: path.join(
          __dirname,
          mockWebRoutesPath ||
            '../create-redwood-app/template/web/src/Routes.tsx'
        ),
      },
      base: path.join(__dirname, '../create-redwood-app/template'),
    }),
  }
})

jest.mock('../../lib/project', () => ({
  isTypeScriptProject: () => false,
}))

// This function checks output matches
const writeFileSyncSpy = jest.fn((_, content) => {
  expect(content).toMatchSnapshot()
})

beforeEach(() => {
  mockWebAppPath = ''
  jest.restoreAllMocks()
  jest.spyOn(fs, 'writeFileSync').mockImplementation(writeFileSyncSpy)
})

describe('Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts', () => {
  it('Matches Auth0 Snapshot', () => {
    addConfigToApp()
    createWebAuth(path.join(__dirname, 'fixtures/dbAuthSetup'), 'auth0', false)
    addConfigToRoutes()
  })

  it('Matches Clerk Snapshot', () => {
    addConfigToApp()
    createWebAuth(path.join(__dirname, 'fixtures/dbAuthSetup'), 'clerk', false)
    addConfigToRoutes()
  })
})

describe('Components with props', () => {
  it('Should add useAuth on the same line for single line components, and separate line for multiline components', () => {
    mockWebAppPath =
      'src/auth/__tests__/fixtures/AppWithCustomRedwoodApolloProvider.js'
    mockWebRoutesPath =
      'src/auth/__tests__/fixtures/RoutesWithCustomRouterProps.tsx'

    addConfigToApp()
    addConfigToRoutes()
  })

  it('Should not add useAuth if one already exists', () => {
    mockWebAppPath =
      'src/auth/__tests__/fixtures/AppWithCustomRedwoodApolloProvider.js'
    mockWebRoutesPath = 'src/auth/__tests__/fixtures/RoutesWithUseAuth.tsx'

    addConfigToApp()
    addConfigToRoutes()
  })
})

describe('Customized App.js', () => {
  it('Should add auth config when using explicit return', () => {
    mockWebAppPath = 'src/auth/__tests__/fixtures/AppWithExplicitReturn.js'

    addConfigToApp()
  })
})

describe('Swapped out GraphQL client', () => {
  let consoleWarn

  beforeAll(() => {
    consoleWarn = console.warn
    console.warn = jest.fn()
  })

  afterAll(() => {
    console.warn = consoleWarn
  })

  it('Should add auth config when app is missing RedwoodApolloProvider', () => {
    mockWebAppPath =
      'src/auth/__tests__/fixtures/AppWithoutRedwoodApolloProvider.js'

    addConfigToApp()

    expect(console.warn).toHaveBeenCalledWith(
      expect.stringMatching(/GraphQL.*useAuth/)
    )
  })
})
