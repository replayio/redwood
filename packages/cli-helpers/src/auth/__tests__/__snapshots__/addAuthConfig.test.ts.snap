// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components with props Should add useAuth on the same line for single line components, and separate line for multiline components 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate="%PageTitle | %AppTitle">
      <AuthProvider>
        <RedwoodApolloProvider useAuth={useAuth} graphQLClientConfig={{ cache }}>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Components with props Should add useAuth on the same line for single line components, and separate line for multiline components 2`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router useAuth={useAuth}
      pageLoadingDelay={400}
      trailingSlashes="always"
      paramTypes={{
        foo: {
          match: /foo/,
          parse: (value: string) => value.split('').reverse().join(''),
        },
      }}
    >
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`Components with props Should not add useAuth if one already exists 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate="%PageTitle | %AppTitle">
      <AuthProvider>
        <RedwoodApolloProvider useAuth={useAuth} graphQLClientConfig={{ cache }}>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Components with props Should not add useAuth if one already exists 2`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router
      pageLoadingDelay={400}
      trailingSlashes="always"
      paramTypes={{
        foo: {
          match: /foo/,
          parse: (value: string) => value.split('').reverse().join(''),
        },
      }}
      useAuth={() => ({
        loading: false,
        isAuthenticated: false
      })}
    >
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`Customized App.js Should add auth config when using explicit return 1`] = `
"import { useEffect } from 'react'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = (props) => {
  const { cache } = props

  useEffect(() => {
    console.log('Running my custom useEffect hook on each render.')
  })

  return (
    <FatalErrorBoundary page={FatalErrorPage}>
      <RedwoodProvider titleTemplate="%PageTitle | %AppTitle">
        <AuthProvider>
          <RedwoodApolloProvider useAuth={useAuth}>
            <AnotherProvider>
              <Routes />
            </AnotherProvider>
          </RedwoodApolloProvider>
        </AuthProvider>
      </RedwoodProvider>
    </FatalErrorBoundary>
  )
}

export default App
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Auth0 Snapshot 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate="%PageTitle | %AppTitle">
      <AuthProvider>
        <RedwoodApolloProvider useAuth={useAuth}>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Auth0 Snapshot 2`] = `
"import { createDbAuth } from '@redwoodjs/auth-providers-web'

export const { AuthProvider, useAuth } = createDbAuth()
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Auth0 Snapshot 3`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router useAuth={useAuth}>
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Clerk Snapshot 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate="%PageTitle | %AppTitle">
      <AuthProvider>
        <RedwoodApolloProvider useAuth={useAuth}>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Clerk Snapshot 2`] = `
"import { createDbAuth } from '@redwoodjs/auth-providers-web'

export const { AuthProvider, useAuth } = createDbAuth()
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Clerk Snapshot 3`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router useAuth={useAuth}>
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`Swapped out GraphQL client Should add auth config when app is missing RedwoodApolloProvider 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const queryClient = {}

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate="%PageTitle | %AppTitle">
      <AuthProvider>
        <QueryClientProvider client={queryClient}>
          <RedwoodReactQueryProvider>
            <Routes />
          </RedwoodReactQueryProvider>
        </QueryClientProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;
